activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -797,6 +835,130 @@ def reconnect_can_restore_state?
+        #
+        # In the latter case, we really ought to guarantee the inner
+        # call will not reconnect (which would interfere with the
+        # still-yielded connection in the outer block), but we currently
+        # provide no special enforcement there.
+        #
+        def with_raw_connection(allow_retry: false, uses_transaction: true)
+          @lock.synchronize do
+            materialize_transactions if uses_transaction
+
+            retries_available = allow_retry ? connection_retries : 0
+            reconnectable = reconnect_can_restore_state?
+
+            if @verified
+              # Cool, we're confident the connection's ready to use. (Note this might have
+              # become true during the above #materialize_transactions.)
+            elsif reconnectable
+              if allow_retry
+                # Not sure about the connection yet, but if anything goes wrong we can
+                # just reconnect and re-run our query
+              else
+                # We can reconnect if needed, but we don't trust the upcoming query to be
+                # safely re-runnable: let's verify the connection to be sure
+                verify!
+              end
+            else
+              # We don't know whether the connection is okay, but it also doesn't matter:
+              # we wouldn't be able to reconnect anyway. We're just going to run our query
+              # and hope for the best.
+            end
+
+            begin
+              result = yield @raw_connection
+              @verified = true
+              result
+            rescue => original_exception
+              translated_exception = translate_exception_class(original_exception, nil, nil)
+
+              if retries_available > 0
+                retries_available -= 1
+
+                if retryable_query_error?(translated_exception)
+                  backoff(connection_retries - retries_available)
+                  retry
+                elsif reconnectable && retryable_connection_error?(translated_exception)
+                  reconnect!(restore_transactions: true)
+                  # Only allowed to reconnect once, because reconnect! has its own retry
+                  # loop
+                  reconnectable = false
+                  retry
+                end
+              end
+
+              raise translated_exception
+            ensure
+              dirty_current_transaction if uses_transaction
+            end
+          end
+        end
+
+        def retryable_connection_error?(exception)
+          exception.is_a?(ConnectionNotEstablished)
+        end
+
+        def retryable_query_error?(exception)
+          # We definitely can't retry if we were inside a transaction that was instantly
+          # rolled back by this error
+          if exception.is_a?(TransactionRollbackError) && savepoint_errors_invalidate_transactions? && open_transactions > 0
+            false
```

Comment:
I don't get this condition. The method returns `nil` by default, so there's no point checking this to finally return `false`. Am I missing something?

Reply:
It was just the way my brain constructed the conditional, special-casing the "if all these things are true, we immediately know it's not retryable" before heading through to the "otherwise, normally" in the `else`... where the equivalent `unless` (which I guess is what you're suggesting?) feels more like it's carving out a subset of the "normally" case. Or something. 

But yeah, different spelling same result -- there's no hidden false vs nil complexity here.

Reply:
Ah, no: `Deadlocked < TransactionRollbackError`. (And it also felt important to call out that _all_ `TransactionRollbackError` are at risk here, in case of future changes to the method.)

Reply:
Ah! My bad then.